lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -25,16 +25,228 @@
+      J.attribute("module_name", memOp.name());
+      J.attribute("depth", (int64_t)memOp.depth());
+      J.attribute("width", (int64_t)width);
+      J.attribute("masked", "true");
+      J.attribute("read", numReadPorts ? "true" : "false");
+      J.attribute("write", numWritePorts ? "true" : "false");
+      J.attribute("readwrite", numReadWritePorts ? "true" : "false");
+      J.attribute("mask_granularity", (int64_t)maskGran);
+      J.attributeArray("extra_ports", [&] {});
+      // Record all the hierarchy names.
+      SmallVector<std::string> hierNames;
+      J.attributeArray("hierarchy", [&] {
+        for (auto p : paths) {
+          const InstanceOp &x = p.front();
+          std::string hierName =
+              x->getParentOfType<FModuleOp>().getName().str();
+          for (InstanceOp inst : p) {
+            hierName = hierName + "." + inst.name().str();
+          }
+          hierNames.push_back(hierName);
+          J.value(hierName);
+        }
+      });
+      // If verification annotation added to the memory op then print the data.
+      if (verifData)
+        J.attributeObject("verification_only_data", [&] {
+          for (auto name : hierNames) {
+            J.attributeObject(name, [&] {
+              for (auto data : verifData) {
+                // Id for the memory verification property.
+                std::string id = data.first.strref().str();
+                // Value for the property.
+                auto verifValAttr = data.second;
+                // Now print the value attribute based on its type.
+                printAttr(J, verifValAttr, id);
+              }
+            });
+          }
+        });
+    });
+  };
+  std::string testBenchJsonBuffer;
+  llvm::raw_string_ostream testBenchOs(testBenchJsonBuffer);
+  llvm::json::OStream testBenchJson(testBenchOs);
+  std::string dutJsonBuffer;
+  llvm::raw_string_ostream dutOs(dutJsonBuffer);
+  llvm::json::OStream dutJson(dutOs);
+  SmallVector<MemOp> dutMems;
+  SmallVector<MemOp> tbMems;
+
+  for (auto mod : circuitOp.getOps<FModuleOp>()) {
+    bool isDut = dutModuleSet.contains(mod);
+    for (auto memOp : mod.getBody().getOps<MemOp>())
+      if (isDut)
+        dutMems.push_back(memOp);
+      else
+        tbMems.push_back(memOp);
+  }
+  std::string seqMemConfStr, tbConfStr;
+  dutJson.array([&] {
+    for (auto memOp : dutMems)
+      createMemMetadata(memOp, dutJson, seqMemConfStr);
+  });
+  testBenchJson.array([&] {
+    for (auto memOp : tbMems)
+      createMemMetadata(memOp, testBenchJson, tbConfStr);
+  });
+
+  auto *context = &getContext();
+  auto builder = OpBuilder::atBlockEnd(circuitOp.getBody());
+  AnnotationSet annos(circuitOp);
+  auto diranno = annos.getAnnotation(metadataDirectoryAttrName);
+  std::string metadataDir = "metadata";
+  if (diranno) {
+    if (auto dir = diranno.get("dirname")) {
+      if (dir.isa<StringAttr>())
+        metadataDir = dir.cast<StringAttr>().getValue().str();
+    }
+  }
```

Comment:
I really think `getAnnotation` should return an `Annotation` instead of a `DictAttr` .. If we don't want to convert the `DictAttr` to an `Annotation`, you can use the `getAs<>` function to make this a bit more succinct. 
```c++
  if (dirAnno) {
    if (auto dir = dirAnno.getAs<StringAttr>("dirname"))
        metadataDir = dir.getValue();
  }
```